### PR TITLE
MRG, DOC: Inverse math

### PR DESCRIPTION
**Pull request by larsoner**
~~Working on making eLORETA have units nAm.~~

~~Along the way~~ Writing out some of the inverse derivations and adding equation links so that they are hopefully easier to follow.

~~This does not work, mostly opening PR so that @agramfort can look at where I've messed up the math :)~~

### CODE REVIEW

File: doc/_includes/inverse.rst [hunk 1]
```diff
@@ -184,32 +192,63 @@ regularizing the channel covariance matrix.
...
+
+    w(t) &= U^T \tilde{x}(t) \\
+         &= U^T C^{-^1/_2} x(t)\ ,
+
+we can see that the expression for the expected current is just
+
+.. math::
+   :label: inv_j_hat_t_svd
+
+    \hat{j}(t) &= R^C V \Gamma w(t) \\
```
Comment (agramfort) [inline doc/_includes/inverse.rst]: I would use R^{^1/_2} as it's what is used above. Not R^C

Comment (larsoner) [reply doc/_includes/inverse.rst]: Yes that was already there and I wasnt sure. Feel free to push that and the other fix as I'm away for some hours. Will get to it later if you don't

File: doc/_includes/inverse.rst [hunk 2]
```diff
@@ -184,32 +192,63 @@ regularizing the channel covariance matrix.
...
-With :math:`w(t) = U^T C^{-^1/_2} x(t)` the expression for the expected current
-is
+.. `reginv` in our code:
 
-.. math::    \hat{j}(t) = R^C V \Gamma w(t) = \sum_k {\bar{v_k} \gamma_k w_k(t)}\ ,
+.. math::
+   :label: inv_gamma_k
+
+    \gamma_k &= \frac{1}{\lambda_k} \frac{\lambda_k^2}{\lambda_k^2 + \lambda^2} \\
+             &= \frac{\lambda_k}{\lambda_k^2 + \lambda^2}\ .
```
[the remark this answers is not in the document]
Comment (larsoner) [reply doc/_includes/inverse.rst]: I think that's true based on the equations above. But I wonder if instead those should have lambda

[the remark this answers is not in the document]
Comment (larsoner) [reply doc/_includes/inverse.rst]: (otherwise I'm not sure where the regularization parameter went, and also I think the lambda is there in the code)

**Comment by larsoner** [comment]
Rendered doc:

https://17326-1301584-gh.circle-artifacts.com/0/dev/overview/implementation.html#the-minimum-norm-current-estimates

File: mne/minimum_norm/tests/test_inverse.py [hunk 3]
```diff
@@ -305,15 +305,17 @@ def test_inverse_operator_channel_ordering(evoked, noise_cov):
...
 def test_localization_bias_fixed(bias_params_fixed, method, lower, upper,
                                  depth):
     """Test inverse localization bias for fixed minimum-norm solvers."""
     evoked, fwd, noise_cov, _, want = bias_params_fixed
     fwd_use = convert_forward_solution(fwd, force_fixed=False)
     inv_fixed = make_inverse_operator(evoked.info, fwd_use, noise_cov,
                                       loose=0., depth=depth)
-    loc = np.abs(apply_inverse(evoked, inv_fixed, lambda2, method).data)
+    loc = np.abs(apply_inverse(evoked, inv_fixed, lambda2, method,
+                               verbose='debug').data)
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_inverse.py]: revert?

File: mne/minimum_norm/tests/test_inverse.py [hunk 4]
```diff
@@ -349,15 +352,16 @@ def test_localization_bias_loose(bias_params_fixed, method, lower, upper,
...
+])
 def test_localization_bias_free(bias_params_free, method, lower, upper,
                                 kwargs, depth):
     """Test inverse localization bias for free minimum-norm solvers."""
     evoked, fwd, noise_cov, _, want = bias_params_free
     inv_free = make_inverse_operator(evoked.info, fwd, noise_cov, loose=1.,
                                      depth=depth)
     loc = apply_inverse(evoked, inv_free, lambda2, method,
-                        pick_ori='vector', **kwargs).data
+                        pick_ori='vector', verbose='debug', **kwargs).data
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_inverse.py]: revert?

File: mne/minimum_norm/tests/test_inverse.py [hunk 5]
```diff
@@ -425,19 +429,19 @@ def test_apply_inverse_operator(evoked):
     stc = apply_inverse(evoked, inverse_operator, lambda2, "sLORETA")
     assert stc.subject == 'sample'
     assert stc.data.min() > 0
-    assert stc.data.max() < 10.0
+    assert 2 < stc.data.max() < 5
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_inverse.py]: why did you need to update these tests?

File: doc/_includes/inverse.rst [hunk 6]
```diff
@@ -45,8 +29,13 @@ the maximum a posteriori (MAP) estimate under the following assumptions:
...
   known covariance matrix. The noise is not correlated over time.
 
+Computing the inverse operator is accomplished using
+:func:`mne.minimum_norm.make_inverse_operator` and
+:func:`mne.minimum_norm.apply_inverse`. The use of these functions is presented
+in the tutorial :ref:`tut-inverse-methods`.
+
 The linear inverse operator
----------------------------
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (agramfort) [inline doc/_includes/inverse.rst]: MNE is not really a Bayesian approach as you just do a MAP on a probabilistic model. 
A Bayesian approach would involves some estimate of the posterior distribution and integrate something against is. I would rephrase `following from the Bayesian approach is given by the :math:`M \times N` matrix` to `following from regularized maximal likelihood of the above probabilistic model is given by the :math:`M \times N` matrix`.

File: doc/_includes/inverse.rst [hunk 7]
```diff
@@ -74,15 +57,26 @@ The expected value of the current amplitudes at time *t* is then given by
 :math:`\hat{j}(t) = Mx(t)`, where :math:`x(t)` is a vector containing the
```
Comment (agramfort) [inline doc/_includes/inverse.rst]: you use M for the linear operator and the dimension. This is very confusing.

Comment (agramfort) [reply doc/_includes/inverse.rst]: One option is to use K for the inverse operator as it's often refered to as the imaging "kernel". Other option is to use Q for the dimension. So Q=P or Q=3P.

File: doc/_includes/inverse.rst [hunk 8]
```diff
@@ -249,14 +303,38 @@ its variance. Noise normalization serves three purposes:
   resulting from the signals coming from a point current source (a current
   dipole) located at a certain point on the cortex.
 
-In practice, noise normalization requires the computation of the diagonal
-elements of the matrix
+In practice, noise normalization is implemented as a division by the square
+root of the estimated variance of each voxel. In computing these noise
+normalization factors, it's convenient to reuse our "weighted eigenleads"
+definition from equation :eq:`inv_j_hat_t` in matrix form as
```
Comment (agramfort) [inline doc/_includes/inverse.rst]: aroudn here `complex estimate current` -> `complex estimated current`

File: doc/_includes/inverse.rst [hunk 9]
```diff
@@ -74,15 +57,26 @@ The expected value of the current amplitudes at time *t* is then given by
...
+.. math::
+   :label: inv_m
 
-.. math::    M = R G^T (G R G^T + \lambda^2 C)^{-1}\ ,
+    M &= R' G^T (G R' G^T + C)^{-1} \\
+      &= R \lambda^{-2} G^T (G R \lambda^{-2} G^T + C)^{-1} \\
+      &= R \lambda^{-2} G^T \lambda^2 (G R G^T + \lambda^2 C)^{-1} \\
+      &= R G^T (G R G^T + \lambda^2 C)^{-1}\ ,
 
 where the unknown current amplitude is now interpreted in terms of the
```
Comment (agramfort) [inline doc/_includes/inverse.rst]: around here I would explicit:

```
\min_j (y - Gj)^\top C^{-1} (y - Gj) + \lambda^2 j^\top R^{-1} j
```

you also see that transpose in latex should be written \top so it does not read like the lettre T that could use elsewhere (eg to denote time index)

Comment (agramfort) [reply doc/_includes/inverse.rst]: I don't know if you have a notation for the data so I used y here.

File: mne/minimum_norm/tests/test_inverse.py (hunk 5 above)
Comment (larsoner) [reply mne/minimum_norm/tests/test_inverse.py]: Didn't need to, but it helps during changes to the `inverse` code to have them be more precise

File: mne/minimum_norm/tests/test_inverse.py (hunk 4 above)
Comment (larsoner) [reply mne/minimum_norm/tests/test_inverse.py]: The `'debug'` shouldn't hurt during runs, and helps when they do fail

File: doc/_includes/inverse.rst (hunk 7 above)
Comment (larsoner) [reply doc/_includes/inverse.rst]: I'll go with the `Q` approach because I think Matti might be partial to `M` for the inverse

**Comment by larsoner** [comment]
@agramfort only https://github.com/mne-tools/mne-python/pull/7160/commits/bf467a20af5d9725e0612104ba69f1847a38afac is new but I rebased + force-pushed, should have addressed all your comments:

https://18038-1301584-gh.circle-artifacts.com/0/dev/overview/implementation.html#the-minimum-norm-current-estimates

**Comment by larsoner** [comment]
FYI I killed Travis because it's going to fail due to failures in `master` (and was passing last time `master` was green)

File: doc/_includes/inverse.rst [hunk 10]
```diff
@@ -104,18 +105,51 @@ discrepancy between the measured and predicted data is tolerable.
...
+
+.. math::
+   :label: inv_tilde_x_t
+
+    \tilde{x}(t) = C^{-^1/_2}x(t)
+
+as the whitened measurement vector at time *t*. The spatial
+whitening operator :math:`C^{-^1/_2}` is obtained with the help of the
+eigenvalue decomposition
+:math:`C = U_C \Lambda_C^2 U_C^\top` as :math:`C^{-^1/_2} = \Lambda_C^{-1} U_C^\top`.
```
[the remark this answers is not in the document]
Comment (larsoner) [reply doc/_includes/inverse.rst]: Perhaps, but I'd rather sort that notation change out as a separate PR. I'm trying as much as possible to stick to the notation in the existing document, and I suspect that changing this will not be trivial.

**Comment by larsoner** [comment]
@hichamjanati I pushed a commit to hopefully address your comments with the exception of the `C^{-^1/_2}` changes, as I think those will require a bit more work to sort out (as a bit of a departure from the existing notation). For example we might want to adopt the "whitener" naming/nomenclature, or we might just want to use the standard definition of `C^{-^1/_2}`, or we might want to show why it doesn't matter which one you use for our purposes. Which is best will take some thinking and math to sort out...

**Comment by larsoner** [comment]
I'll go ahead and merge this once CIs are happy.

BTW @hichamjanati feel free to take a shot at the whitener changes if you want (after merge) :)
